File: java/00078_subsets/Solution.java
import java.util.ArrayList;
import java.util.List;
/**
 * Given an integer array nums of unique elements, return all possible subsets (the power set). The solution set must
 * not contain duplicate subsets. Return the solution in any order.
 */
class Solution {

    /**
     * Generate all possible subsets of numbers from a given set of integers.
     *
     * The function, subsets(), uses the concept of bit manipulation
     * to generate possible subsets.
     *
     * For every integer 'i' from 0 to 2^n, it treats the binary representation of 'i'
     * as a mask for whether to include a number in the subset. Since there are 2^n possible
     * subsets of a set with 'n' numbers, the function goes from 0 to 2^n
     * (exclusive) to generate all subsets.
     *
     * If the jth bit of 'i' is set, it means the jth number of the set is included
     * in the subset, and if it's not set, then the jth number is not in the subset.
     *
     * @param nums:  Array containing the original set of integers.
     * @return The list of all possible subsets. Each subset is represented as a List of Integers.
     */
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        int numSize = nums.length;
        int numSubset = 1 << numSize; // Equivalent to 2^numSize

        // Iterate over all potential subsets
        for (int i = 0; i < numSubset; i++) {
            List<Integer> temp = new ArrayList<>();

            // Check each bit from the current mask
            for (int j = 0; j < numSize; j++) {
                // If the bit is set, add the number corresponding to the bit to the subset
                if ((i >> j & 1) == 1) {
                    temp.add(nums[j]);
                }
            }
            output.add(temp); // Add the current subset to the output
        }
        return output; // Return all subsets
    }
}

File: cpp/00078_subsets/Solution.cpp
#include <vector>
#include <cmath>

/**
 * Given an integer array nums of unique elements, return all possible subsets (the power set).
 * The solution set must not contain duplicate subsets. Return the solution in any order.
 *
 * @class Solution
 * @brief A class that implements functionality to generate all possible subsets from a set of integers.
 */
class Solution {
public:
    /**
     * @brief This function generates all possible subsets from a set of integers.
     *
     * @param nums The original set of integers for which subsets are to be found.
     * @return A vector of vectors where each vector represents a subset of the input vector.
     *
     * The function utilizes bit manipulation to generate the subsets.
     * Each subset is represented by a binary number where a bit being set (1) means the corresponding number is included in the subset,
     * and the bit being unset (0) meaning it's not included in the subset.
     * Therefore, the total number of subsets is equal to 2^n where n is the size of the input set.
     */
    std::vector<std::vector<int>> subsets(std::vector<int>& nums) {
        std::vector<std::vector<int>> output; // A vector to store all the subsets
        int numSize = nums.size(); // Size of the input set
        int numSubset = std::pow(2, numSize); // Number of subsets

        // Loop over the range of total possible subsets
        for (int i = 0; i < numSubset; i++) {
            std::vector<int> temp; // Vector to hold the current subset

            // Loop over the range of input set
            for (int j = 0; j < numSize; j++) {
                // If the bit in the current subset is set add the corresponding number to the subset
                if ((i >> j) & 1) {
                    temp.push_back(nums[j]);
                }
            }
            output.push_back(temp); // Add the subset to the output
        }
        return output; // Return the output
    }
};